wgpu-core/src/device/global.rs

```diff
@@ -2202,17 +2101,14 @@ impl<G: GlobalIdentityHandlerFactory> Global<G> {
                 }
             }
 
-            if let Some(present) = surface.presentation.take() {
+            if let Some(present) = surface.presentation.lock().take() {
```

Comment:
You should acquire the lock once, and then perform both the check and the update while it's held, to avoid race conditions. Suggest:

```
            {
                let mut presentation_guard = surface.presentation.lock();
                if let Some(present::Presentation { acquired_texture: Some(_), .. }) = *presentation_guard {
                    break E::PreviousOutputExists;
                }

                *presentation_guard = Some(present::Presentation {
                    device_id: id::Valid(device_id),
                    config: config.clone(),
                    num_frames,
                    acquired_texture: None,
                });
            }
```